packages/core/src/temporal/NativeJsTemporal.js

```diff
@@ -1,105 +1,92 @@
 /*
- * @copyright (c) 2016, Philipp Thürwächter & Pattrick Hüper
+ * @copyright (c) 2022, Philipp Thürwächter & Pattrick Hüper & Michał Sobkiewicz
  * @license BSD-3-Clause (see LICENSE in the root directory of this source tree)
  */
 
-import { assert, requireNonNull } from '../assert';
-import { UnsupportedTemporalTypeException } from '../errors';
-
-import { Instant } from '../Instant';
-import { LocalDate } from '../LocalDate';
-import { LocalTime } from '../LocalTime';
-import { MathUtil } from '../MathUtil';
+import { requireInstance, requireNonNull } from '../assert';
+import { IllegalArgumentException } from '../errors';
+import { Instant } from '../js-joda';
 import { ZoneId } from '../ZoneId';
-
-import { ChronoField } from './ChronoField';
-import { TemporalQueries } from './TemporalQueries';
 import { TemporalAccessor } from './TemporalAccessor';
+import { TemporalField } from './TemporalField';
+import { TemporalQuery } from './TemporalQuery';
 
 /**
- * A wrapper around a native javascript Date instance that
- * implements TemporalAccessor functionality
+ * A wrapper around a javascript Date or a moment instance that implements TemporalAccessor functionality.
  */
 class NativeJsTemporal extends TemporalAccessor {
-
     /**
      * @param {!(Date|moment)} date - a javascript Date or a moment instance
-     * @param {ZoneId} [zone=ZoneId.systemDefault()] - the zone of the temporal, defaults to ZoneId.systemDefault()
+     * @param {ZoneId} [zoneId = ZoneId.systemDefault()] - the zone of the temporal, defaults to ZoneId.systemDefault()
      * @private
      */
-    constructor(date, zone=ZoneId.systemDefault()){
+    constructor(date, zone = ZoneId.systemDefault()) {
         super();
-        this._zone = zone;
-        if(date instanceof Date) {
-            this._epochMilli = date.getTime();
-            return;
-        } else if(typeof date.toDate === 'function' &&  date.toDate() instanceof Date) {
-            // it's a moment
-            this._epochMilli = date.toDate().getTime();
-            return;
+        this._date = requireNonNull(date, 'date');
+        this._zone = requireNonNull(zone, 'zone');
+        switch (date.constructor.name) {
+            case 'Date':
+                this._delegate = () => Instant.ofEpochMilli(date.getTime()).atZone(zone);
```

Comment:
In that case `NativeJsTemporal` is not a wrapper. I'll update esdoc and remove mutability.

Reply:
true 👍

Reply:
How about removing `NativeJsTemporal` and returning `TemporalAccessor` (or even `ZonedDateTime`?) directly from `nativeJs()`? I don't think that implementing all `TemporalAccessor` methods just to delegate them adds much value now. What do you think?

Reply:
so you mean some static function  like `nativeJsToZonedDateTime(date: Date | Moment):ZonedDateTime`.

Yes, sounds reasonable, but we should keep the old `nativeJs` for backwards compatibility and mark it as deprecated.

Reply:
I mean `nativeJs` doesn't have to return `NativeJsTemporal` (which is not public anyway!), but can simply return `TemporalAccessor` (to be abstract) or even `ZonedDateTime` (which is less flexible but gives more guarantees). I think I would prefer first option as there is always room for change later if really needed.

Reply:
Something like:

```javascript
/**
 * @param {!(Date|moment)} date - a javascript Date or a moment instance
 * @param {ZoneId} [zone = ZoneId.systemDefault()] - the zone of the temporal, defaults to ZoneId.systemDefault()
 * @returns {TemporalAccessor}
 */
export function nativeJs(date, zone) {
    requireNonNull(date, 'date');
    requireNonNull(zone, 'zone');
    switch (date.constructor.name) {
        case 'Date':
            return Instant.ofEpochMilli(date.getTime()).atZone(zone);
        case 'Moment':
            return Instant.ofEpochMilli(date.valueOf()).atZone(zone);
        default:
            throw new IllegalArgumentException('date must be either a javascript date or a moment');
    }
}
```

Reply:
But returning `ZonedDateTime` would be a breaking change correct ? In that case we should introduce a new function. 
For the first option, i think its fine to kickout `NativeJsTemporal`

Reply:
> But returning `ZonedDateTime` would be a breaking change correct ? In that case we should introduce a new function.

I wouldn't say so. `NativeJsTemporal` was used just as an almost anonymous implementation of `TemporalAccessor`, didn't provide any additional methods. It wasn't public, so no one should be able to use it with some static checks in place. Even if `NativeJsTemporal` could be imported in some way, why would anyone like to test for `instanceof NativeJsTemporal` where `instanceof TemporalAccessor` could be used instead? I can't really imagine any other breaking statement.

IMO this is perfectly valid and fully compatible substitution.

Reply:
I mean the contract of `nativeJs` seems to be that it returns `TemporalAccessor` which will be fulfilled even more than before ;)

Reply:
You are right, `ZonedDateTime` is implementing `TemporalAccessor`, so the contract did not change.

Fine for me 👍

Reply:
How about changing return type to `Temporal` to provide even more guarantees to the users? My biggest concern is that `ZonedDateTime`, apart from `TemporalField`s, supports `TemporalUnit`s. Users can mistakenly start using `nativeJs(date).isSupported(TemporalUnit)` without realizing that this is not officially supported. It is safer to throw an `IllegalArgumentException` instead of returning true which I did in my rewritten version of `NativeJsTemporal`. On the other hand, I think it would be convenient for everyone to be able to use `Temporal` methods which are supported anyway at no cost, but this should be stated explicitly in return type.

Reply:
Still, contract doesn't change, as `Temporal` extends `TemporalAccessor`.

Reply:
I've made decision by myself. `nativeJs` guarantees that it returns `ZonedDateTime` and I think that this is the most useful option, as callers know exactly which fields and units are going to be supported safely.

Reply:
IMO its fine to return ZonedDateTime instead of Temporal, because thats what the function is technically doing.

But after thinking about it, a better design of the API would be to have something like `nativeToInstant(date)` and `nativeToZonedDateTime(date, zone)` because that express what the functions are doing. 

But your PR is already an improvement and extend the existing functionality without breaking something. I just wonder if there are edge case where the new implementation behaves different then the old one.

Reply:
Then you would probably need to provide `nativeToLocaldate(date, zone)`, `nativeToOffsetTime(date, zone)` and so on...

Reply:
The best API I can imagine would be just `nativeToInstant(date)` and nothing more. `ZonedDateTime` could be easily obtained with `nativeToInstant(date).atZone(zone)`. Same would be true for most of other temporals with `.ofInstant(instant, zone)`.

Reply:
Ack, i would suggest to rewrite nativejs as proposed, but mark as deprecated and at a new function nativeToInstant

Reply:
If you are liking to make such big changes, what would you say for replacing both `nativeJs` and `convert` with the following 2 functions:

```typescript
dateFromInstant(Instant): Date
dateToInstant(Date): Instant
```
?

Naming I propose is almost 1:1 based on:
- https://docs.oracle.com/javase/8/docs/api/java/util/Date.html#from-java.time.Instant- (`dateFromInstant`)
- https://docs.oracle.com/javase/8/docs/api/java/util/Date.html#toInstant-- (`dateToInstant`)

I believe it is much easier to remember than `nativeJs` and `convert`.

For completness, I would also be happy to add:
```typescript
momentFromInstant(Instant): Moment
momentToInstant(Moment): Instant
```

Reply:
My bad. `convert` supports more than just `Instant`, so it wouldn't be direct replacement. Having some `localDate` for example, one would have to invoke `dateFromInstant(Instant.from(localDate.atStartOfDay(zone)))`. On the other hand this makes clear, what time value is going be used and so on.

I would personally stick to simplified view `Instant` ≈ `Date` ≈ "kind of timestamp". This way it doesn't feel unnatural, that it's impossible to convert `LocalDate` to `Date` as it is also not possible to convert from `LocalDate` to `Instant` without providing some additional data.

Reply:
I withdraw from

```typescript
momentFromInstant(Instant): Moment
momentToInstant(Moment): Instant
```

as it's very easy to convert between `Moment and `Instant` like so:
- `moment(dateFromInstant(instant))`,
- `dateToInstant(moment.toDate())`.

As for the:
```typescript
dateFromInstant(Instant): Date
dateToInstant(Date): Instant
```
i sustain my proposal.